REFRAME.md

```diff
@@ -0,0 +1,291 @@
+    # However, we can use this to describe that we have a union keyed by numbers rather than strings.
+    type TransferProtocol union {
+        | Bitswap
+        | GraphSync-FILv1
+    } representation structured 
+    
+    type Bitswap struct {
+        ID SingletonInt(0x0900) # https://github.com/multiformats/multicodec/blob/f5dd49f35b26b447aa380e9a491f195fd49d912c/table.csv#L133
+    }
+    
+    type GraphSync-FILv1 struct {
+        ID SingletonInt(0x0910) # https://github.com/multiformats/multicodec/blob/f5dd49f35b26b447aa380e9a491f195fd49d912c/table.csv#L134
+        PieceCID Link  # Taken from: https://github.com/filecoin-project/index-provider/blob/305bfab501d8850a5b6761df7af6c38ba2359a85/metadata/graphsync_filecoinv1.ipldsch
+        VerifiedDeal Bool
+        FastRetrieval Bool
+    }
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"FindProviders" : {
+    "Key" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"FindProviders" : {
+    "Key" : {"/":{"bytes":"AXIUBPnagss"}},
+    "Providers" : [
+        {"Node":{"Peer":{"ID":{
+            "/":{"bytes":"EncodedPeerID"}},
+            "Multiaddresses":[{"/":{"bytes":"EncodedAddr1"}}, {"/":{"bytes":"EncodedAddr2"}}]}}},
+        {"Node":{"Peer":{"ID":{
+            "/":{"bytes":"EncodedPeerID2"}},
+            "Multiaddresses":[{"/":{"bytes":"EncodedAddr1"}}, {"/":{"bytes":"EncodedAddr2"}}]}}},
+                "Proto" : [{
+                    "ID": 2320, # the integer of the graphsync-filv1 code
+                    "PieceCID" : {"/": "bsome-base32-CidV1"},
+                    "VerifiedDeal" : true,
+                    "FastRetrieval" : false,
+                },
+                "Proto" : {
+                    "ID": 2304 # integer bitswap code
+                }]
+    ]
+}}
+```
+
+#### GetIPNS
+
+A message for finding the latest IPNS records for a given identifier.
+
+```ipldsch
+    type GetIPNSRequest struct {
+        ID Bytes # An IPNS Identifier, matches a peerID https://github.com/libp2p/specs/blob/f433ad595224cf33d916c166d1738f11aadfa9f7/peer-ids/peer-ids.md
+    }
+```
+
+```ipldsch
+    type GetIPNSResponse struct {
+        Record Bytes # An IPNS record
+    }
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"GetIPNS" : {
+    "ID" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"GetIPNS" : {
+    "Record" : {"/":{"bytes":"firstRecord"}}
+}}{"GetIPNS" : {
+    "Record" : {"/":{"bytes":"laterRecord"}}
+}}...
+```
+
+#### PutIPNS
+
+A message for putting the latest IPNS records for a given identifier.
+
+```ipldsch
+    type PutIPNSRequest struct {
+        ID Bytes # IPNS Identifier
+        Record Bytes # IPNS record
+    }
+    
+    type PutIPNSResponse struct {}
+```
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"PutIPNSRequest" : {
+    "ID" : {"/":{"bytes":"AXIUBPnagss"}},
+    "Record" : {"/":{"bytes":"AXIUBPnagss"}}
+}}
+```
+
+Response:
+```
+{"PutIPNSResponse : {}"}

```

Comment:
I think for now this is ok. It might be that in the future we want to return other things like if the server knows about a more recent/better IPNS record (which is what the DHT code does), however that kind of situation also seems like something that'd be a reasonable error (i.e. indicating that the put is too old to be valid and that you can do a Get if you want the latest value).